### PR TITLE
New Storage API + File FieldType

### PR DESCRIPTION
**Pull request by JedWatson**
**WIP For Review and Feedback**

I've been working on a new, generic and extendable Storage API for Keystone and corresponding File FieldType for 0.4 to replace the previous mess of implementations we had with LocalFile, AzureFile, S3File etc.

This PR should be enough to get feedback and check some design decisions I've made - if you're interested in using the File and Storage functionality in Keystone please chime in!

## Background

We need to move away from providing individual field types for different storage providers. There's too much code in each, they're poorly structured and add too much overhead to developing and supporting Keystone. They are, in fact, the largest remaining block to getting 0.4 released with the field types rewrite and new API.

My proposal is to provide a new `keystone.Storage` class that you can provide an `adapter` to, where the adapter can be an external package from npm and handles file uploading / downloading / deleting / etc. as needed by implementing the specified API for Storage Adapters.

We then will create four generic field types in Keystone for handling files that expose a common interface (react components) and require a Storage instance to be provided in the field options. The generic fields will be:

* File
* Files
* Image
* Images

(as the Image field types are new, they aren't necessarily required for 0.4 launch, but planning for them will remove a fair bit of the complexity that crept into the file fields over the course of the 0.2 and 0.3 releases)

The first Adapter I've written is `FS`, we'd need to add `Azure` and `S3` to release 0.4 without losing any existing functionality.

The new File field and Storage API are, at this stage, quite "bare bones". There's a lot you can do (and may need to do) with files in node.js applications and the goal is not to support everything, but to provide a sufficient default set of functionality out of the box that doesn't get more complex as we add new providers.

This means that some features (value formatting, hooks, etc) have been lost in the transition but IMO the implementations were too complex to follow, poorly documented, and inconsistent / buggy. So I've taken it back to the drawing board and hopefully come up with a more elegant / scalable solution at the cost of some functionality.

We can introduce more functional requirements than I've spec'd so far, and delay 0.4, or expand the spec and implementation over time. One of my goals in opening this up for review is to get feedback on how well I've done in terms of the design and initial implementation, and to find out whether anybody has any serious concerns about losing functionality I haven't brought forward.

The implementation isn't complete yet because I wanted to open it up to feedback ASAP; this should be reviewed in combination with the implementation in the [keystone-test-project](https://github.com/keystonejs/keystone-test-project/tree/file-fields) which actually uses the new Storage API to create a [File](https://github.com/keystonejs/keystone-test-project/blob/file-fields/models/File.js) list and has the ability to test validation and uploading in its [API Test UI](https://github.com/keystonejs/keystone-test-project/blob/file-fields/tests/api-react/file/1-File.js).

Files aren't actually written to disk yet, and you can't update values through the Admin UI Edit Form, but otherwise it's complete in terms of spec and I've documented the implementation so you can follow what's going on.

To get this ready for release we'll need to write docs, write the spec for the `Storage API` and `Adapters`, write a spec for the `File` FieldType and complete the `FS` adapter.

## Storage API

See the [implementation](https://github.com/keystonejs/keystone/blob/storage/lib/storage/index.js)

`Storage` instances provide the schema for the `File` field to store in the database. By default, it looks like this:

```javascript
{
	filename: String, // the name of the file
	filesize: Number, // the size of the file in bytes
	filetype: String, // the mime type of the file
	url: String, // optional public URL of the file
}
```

Adapters can override the schema but need to conform to a few standards (I think the minimum is storing the name of the file in a `filename` path)

`Storage` instances provide `uploadFile` and `removeFile` methods; I think that's all we need for v1 but let's discuss.

The result returned from the `uploadFile` method is the value actually stored in the field schema in the database; this is neat because it leaves Adapters almost entirely in control over what's going on both in the implementation and stored data as long as the generic UIs for column, filters and editing are sufficient. We can extend this over time as we make the components richer.

### Usage

Create a new Storage instance like this:

```javascript
var UploadStore = new keystone.Storage({
	adapter: keystone.Storage.Adapters.FS,
	path: 'uploads',
	publicPath: '/files/uploads/',
});
```

In this case (for the FS adapter) you need to provide a `path` option (which is where the files are stored on disk) and optionally a `publicPath` option (which is prepended to the file name to generate the publicly accessible URL for the file).

## Adapters

Adapters need to provide `getFileURL`, `uploadFile` and `removeFile` methods.

See the WIP [FS Adapter implementation](https://github.com/keystonejs/keystone/blob/storage/lib/storage/adapters/fs/index.js)

## File Field

I've created a draft implementation of the File field that uses the new Storage API. You configure the field like this:

```javascript
myfile: { type: Types.File, storage: UploadStore, required: true, initial: true },
```

The field exposes four methods (which are also underscore methods) -

`format(item)` - returns the filename or an empty string
`upload(item, file, callback)` - uploads a new file and stores the result in the field
`reset(item)` - resets the field value
`remove(item)` - deletes the stored file and resets the field value

I'm still working out whether we need to support additional _special_ paths.

Previously we had a path for action (e.g. `delete` / `reset`) which I think we can get around by detecting strings passed as the field value (e.g. `''` for reset, `'!delete'` for delete) - these will never be interpreted as values to upload, which seems safe and simpler.

We also had a path for uploading but it seems fine to detect a file that needs to be uploaded in the field value, which has been successful with the `CloudinaryImageType`.

## Testing

Clone the [Test Project](https://github.com/keystonejs/keystone-test-project) and switch to the `file-fields` branch. Make sure you have a local clone of Keystone with the `storage` branch checked out (run `npm install`, there are package changes) and run `npm link` in Keystone then `npm link keystone` in the test project to set it all up. Finally, run `npm start` in the test project.

See [models/File](https://github.com/keystonejs/keystone-test-project/blob/master/models/File.js) for the Storage implementation.

[Sign into Keystone](http://localhost:3000/keystone) and open [localhost:3000/api](http://localhost:3000/api) to use the UI to test validation and uploading files.

We'll add more functionality to this test harness so everything can be explicitly tested (e.g. resetting values, deleting files).

Once you've created an item and uploaded a file, you can view it in the Admin UI, although the components are still WIP and you can't edit file fields there yet.

## Breaking Changes

This is a pretty significant breaking change for anyone using LocalFile / Azure / S3 fields in Keystone, including changes to the database schema.

We're losing the `originalname` and `path` data from the database (which are redundant anyway), we're changing paths (`size` -> `filesize`) and we're losing the `exists` and `href` virtuals. Exists is kind of useless and I don't think it was a good thing to add in the first place, but we could restore it if we want to; `href` has been replaced by the consistent `url` implementation that's now actually persisted to the database.

The other major design decision to call out is the removal of the `path` option from the fields themselves. That's now set as a `Storage` option. This means we don't need to store path information in the database (because it would always be consistent) and if you ever move files on disk you just need to change the Storage options. If anything happened to change the public URLs of stored files, you'd need to write an update script to manage it.

If you have different fields that should store files at different paths, just create multiple storage instances with the same adapter. They're cheap and it's clearer that way.

And finally we're not actually registering the storage instances on the Keystone instance itself; I don't think it's necessary. If you want access to the Storage instance outside of the field functionality (which is a pretty valid use case) or want to share it across multiple lists, just create it in a separate file and require it as needed.

### CODE REVIEW

**Comment by creynders** [comment]
`originalName` is not redundant. It should contain thefile name as it was before upload and potential renaming.

**Comment by JedWatson** [comment]
@creynders I more meant path is redundant (because now the adapter provides the path), but in the default `FS` adapter `originalname` is used to populate the `filename` anyway - so you'd just have the same value in both fields.

I haven't implemented an option to change the stored filename yet (we had options like that before) so we could pretty easily add that in, and include the option to retain the `originalname` as adapter config.

Not sure if that's best dealt with at the `Storage` or `Adapter` level; the adapter provides the schema so even if we never built `originalname` into keystone it would be trivial to add support for in a project.

**Comment by creynders** [comment]
Sounds good! Sorry again for being a jerk! Wasn't thinking. So to be official: awesome job! You finally cracked the FFF :)

**Comment by wmertens** [comment]
I would really like to see generic array fields replace the specific array field types though…

I recently implemented a file download endpoint that gets the list, the item id and the filename, and uses that to determine if the user has access before sending it with the original filename.

Should this proposal also concern itself with access control? Perhaps a uuid per file and standard callbacks?

**Comment by jstockwin** [comment]
I'm sure you're aware of this, but we should also well document the breaking changes. Specifically, if I have some e.g. S3 Files already live on a site, how do I make them play nice with the new field without having to re-upload everything.

**Comment by JedWatson** [comment]
@jstockwin of course 👍 

I'm thinking we not just document the changes but publish reference code to do the database migration so file fields can be upgraded without issues or everyone working out how to implement it themselves.

**Comment by creynders** [comment]
Ok, just reviewed in full. LGTM so far! It's definitely a good idea you went bare bones. I failed due to trying to rewrite everything while maintaining backwards compatibility _and_ all functionality, which is impossible due to the inconsistencies. 

Two things:
- I agree with @wmertens regarding the array fields. I _really_ think we shouldn't have separate fields for arrays. I'd prefer having a `multi` option which allows storing any kind of value in arrays.
- I'm not entirely convinced it's a good idea to drop `path`. In my experience with a bunch of CMS's I've had some serious difficulties with dynamic path resolution regarding maintainability, debugging, etc. I.e. it looks good on paper, but makes things hard in production.

**Comment by wmertens** [comment]
In 0.3 there was a `format` feature. This could be reinstated as a react pure function:

```
format({React, filename, type, foo}) {<div>I have {filename}</div}
```

adding React to the properties makes it use the React from the Keystone UI.

**Comment by mxstbr** [comment]
I think we might need to do that with the general extensibility though, why would we do this specifically for the file fields only?

**Comment by wmertens** [comment]
No, this was just an example for files. Perhaps we can call the setting Field, and it gets the current input field parts, so it can prepend, append, interject and replace UI?

The default Field would then be `const Field = ({React, Input, fieldProps}) => <Input {...fieldProps}/>` or something along those lines, and if you override it yours gets used instead. Different field types would have specific extra UI that is overridable, passed via the `fieldProps`, for example file types could have the Thumbnail component be overridable separately.

So basically provide all the things needed to easily tweak the UI without having to copy the entire input code for a tiny change.

Would be cool if we also allow it on labels or something so you can add buttons to a section.

**Comment by josephg** [comment]
@JedWatson @molomby and I spent a few hours nutting out how this should work over the last few days.

My big insight with all this is that there's two quite different use cases, which are similar but have a big influence on what APIs you're after:

1. **Keystone is a blob store**. All communication to the blob store happens through keystone, and you think of the backing container as a magicish black box to put files in. If the files are stored on disk, the filename of the chunk is totally irrelevant. (You might want to display the original name at some point to the user, but that string has no semantic meaning.) If the mongo document is deleted, the file should be deleted too. File names on disk will be either a hash of their content or a random string.
2. **Keystone describes some files you already have**. You might have a directory of static assets and you want to use keystone's file APIs to upload new versions of the files. The important distinction here is that the files have a meaning outside of keystone. Keystone stores *references* to the files on disk. If the database reference to the file is removed from keystone, you may or may not want to delete the actual file on disk.

The code is reasonably similar for both use cases, but the semantics change.

The API will probably be something like this:

```javascript
var storage = new keystone.Storage({
	adapter: keystone.Storage.Adapters.localFileStore,
	fsOptions: {
		path: keystone.expandPath('./uploads'),
	},
	getFilename: (file, i) => `${inflect.slug(file.originalName)}}`, // Can take a 3rd callback argument
	onConflict: 'retry', // 'retry' | 'overwrite' | 'error'
	publicUrlBase: '/public/uploads'
});

File.add({
	name: { type: String },
	file: { type: Types.File, storage: storage, required: true, initial: true },
});
```

@JedWatson thinks for now we should only support all the files living in a single directory. But that should be reasonably easy to change later in the filesystem adapter.

Other adapters we might want:

```javascript
	adapter: keystone.Storage.Adapters.S3,
	s3: {
		path: 'keystone-uploads',
		bucket: 'my-bucket',
		accessKey: 'ABASdfasdfasd',
		secretAccessKey: 'asdfasdfi87a6sdf9',
		policy: 'ACL_PRIVATE',
		region: 'ap-southeast-2',
	}),
	getFilename: (file, i) => `${inflect.slug(file.originalName)}}`, // same as above. Defaults to not giving S3 a filename.
	onConflict: 'retry', // 'retry' | 'overwrite' | 'error'

	// ... or....
	adapter: keystone.Storage.Adapters.inlineBlobStore, // Store the blobs as base64 strings in the primary keystone database
	adapter: keystone.Storage.Adapters.mongoBlobStore({ db: 'asdf', collection: 'asdf'. /* .. */ }),
	adapter: keystone.Storage.Adapters.inMemory({}), // for testing
	adapter: keystone.Storage.Adapters.redisBlobStore({ host: 'localhost', port: 6379, dbNumber: 0 }),
```

Adapters have a set of functions something like this:

```
getPublicUrl(id) >> url
upload(file) >> {id, metadata}
download(id) >> Stream
remove(id)
getMetadata(id) >> metadata
```

We're not sure whether the adapter gets to store arbitrary fields in the document. I don't like the extra complexity from having fields that the adapter cares about appear and disappear based on the whim of the user's schema. And the adaptor methods should refer to the file / blob based on an adaptor-specified identifier.

---

We had a long conversation about how adapters should be initialized. Jed thinks they should look like this:

```javascript
var storage = new keystone.Storage({
	adapter: keystone.Storage.Adapters.localFileStore,
	fsOptions: {
		path: keystone.expandPath('./uploads'),
	},
});
```

I think they should look like this:

```javascript
var storage = new keystone.Storage({
	adapter: new keystone.Storage.Adapters.localFileStore({
		path: keystone.expandPath('./uploads'),
	},
});
```

Jed's argument is that keystone can share / interpret some of the adapter configuration the same way it works with fields. I think adapters are fundamentally a bag of functions, and how they are initialized is up to the user. If adapters want to share code they should do it by calling common utility functions.

---

We also talked about whether there should be an FS adapter wrapper class (`Storage`).

- Jed likes the idea because he wants to be able to expose a nice api for, say, REST endpoints to upload files.
- I think thats the job of the adapter - the adapter already has API endpoints for uploading files, and you should just call your adapter 's `uploadFile` function yourself. There's fundamentally two views / interactions here: interactions with blobs in the blob store, and interactions with database entries (which describe blobs in the blob store). I'd like those to be represented by two APIs (the adapter and mongoose, respectively). Not three (the adapter, storage and mongoose). But if we're sharing a bunch of code in adapters via a wrapper keystone makes anyway, then you probably don't want to interact with the adapter directly anyway.

---

I'm going to update the PR / branch to include the changes then merge it into master as soon as we can.

**Comment by josephg** [comment]
@JedWatson Ready for review.

**Comment by JedWatson** [comment]
@josephg would you mind resolving the conflicts so this is ready to merge please?

**Comment by jstockwin** [comment]
Can we make sure to document the changes and have some example code for fixing the breaking change in live apps _before_ this gets merged? Or will that only happen just before 0.4 goes live to npm?

**Comment by JedWatson** [comment]
@jstockwin things won't break for anyone when this gets merged, because it's all additional. It's when we pull out the `LocalFile`/`s`, S3 and Azure stuff that the breaking change will take effect. I'm not planning to do that immediately.

But yes, this will be comprehensively documented and feature in the upgrading guide from 0.3 > 0.4

File: fields/types/number/NumberFilter.js [hunk 1]
```diff
@@ -10,92 +10,108 @@ const MODE_OPTIONS = [
...
-			minValue: '',
-			maxValue: '',
+			filter: getDefaultValue(),
 		};
 	},
 
 	componentDidMount () {
 		// focus the text input
-		ReactDOM.findDOMNode(this.refs.input).focus();
+		findDOMNode(this.refs.focusTarget).focus();
```
Comment (wmertens) [inline fields/types/number/NumberFilter.js]: * Why do we use string refs in new code?
* With React 15, findDOMNode is not needed on DOM refs, they already are the DOM node
* Would be nicer to put this in a `focusSelf()` method since it's used twice

File: fields/types/number/NumberFilter.js [hunk 2]
```diff
@@ -10,92 +10,108 @@ const MODE_OPTIONS = [
...
 	handleChangeBuilder (type) {
 		const self = this;
 		return function handleChange (e) {
-			const { value } = e.target;
-			const { modeValue } = self.state;
-			const { onChange } = self.props;
-			self.setState({
-				[type]: value,
-			});
+			const { filter, onChange } = self.props;
```
Comment (wmertens) [inline fields/types/number/NumberFilter.js]: I rather liked that `value` instead of having to refer to it fully each time?

File: lib/storage/adapters/fs/index.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		}
+		throw e;
+	}
+}
+
+function FSAdapter (options, schema) {
+	if (!schema.filename) throw Error('Cannot use FSAdapter without storing filename');
+
+	this.options = options.fs || {};
+	console.log('FSAdapter', this.options);
```
Comment (wmertens) [inline lib/storage/adapters/fs/index.js]: Could we start using https://github.com/visionmedia/debug ? Lots of frameworks use it, including Mongoose and Express, and it makes the debugging experience much nicer.

File: lib/storage/index.js [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+};
+
+/**
+	This is the default set of schema fields that are enabled. The storage
+	options can override these (must be supported by the adapter, or the
+	adapter must use the default schema implementation)
+*/
+var SCHEMA_FIELD_DEFAULTS = {
+	// filename: true,
+	size: true,
```
Comment (wmertens) [inline lib/storage/index.js]: How about requiring a hash, like md5? That way, etags can be generated and verified without hitting the storage adapter.

File: lib/storage/adapters/fs/index.js (hunk 3 above)
Comment (webteckie) [inline lib/storage/adapters/fs/index.js]: fyi, visionmedia debug is already used by keystone.

File: fields/types/boolean/test/explorer.js [hunk 5]
```diff
@@ -1,8 +1,10 @@
 module.exports = {
 	Field: require('../BooleanField'),
 	Filter: require('../BooleanFilter'),
+	readme: require('fs').readFileSync('./fields/types/boolean/Readme.md', 'utf8'),
```
Comment (wmertens) [inline fields/types/boolean/test/explorer.js]: I understand this uses `brfs` to inline the code, but this way of doing it makes the dependency on the readme file opaque (and relies heavily on browserify if I understand correctly).

Why not import the file and define a loader that reads the `.md` files as a string?

Comment (JedWatson) [inline fields/types/boolean/test/explorer.js]: @wmertens what's that look like?

File: fields/types/number/NumberFilter.js (hunk 1 above)
Comment (JedWatson) [inline fields/types/number/NumberFilter.js]: :+1: to all of that. it really bothers me that this stuff all got caught up with this PR though :(

I'm really tempted to merge into master and fix there before they get any more divergent, since this is already kind of a mess of a PR

File: fields/types/number/NumberFilter.js (hunk 2 above)
Comment (JedWatson) [inline fields/types/number/NumberFilter.js]: I'm OK with changing that back, any reason you prefer it this way @jossmac?

File: fields/types/boolean/test/explorer.js (hunk 5 above)
Comment (wmertens) [inline fields/types/boolean/test/explorer.js]: Well, for webpack you define the loader for `.md` files to be `raw-loader`, and then `import readme from '../Readme.md'` will do what you want.

For browserify, looks like you need to add https://github.com/JohnPostlethwait/stringify

File: lib/storage/adapters/fs/index.js (hunk 3 above)
Comment (JedWatson) [inline lib/storage/adapters/fs/index.js]: Yes, we should be adding useful `debug` calls. I'm not sure that these qualify though, this is really in-development inspection of what's going on..? I was just about to remove them.

File: lib/storage/index.js (hunk 4 above)
Comment (JedWatson) [inline lib/storage/index.js]: I'm down with adding that to the schema. Keen to get this merged with the current feature set, @wmertens are you interested in adding that?

**Comment by webteckie** [comment]
In the keystone-test-project reference implementation, what is the following doing?
```
transform.toJSON(File);
```

File: lib/storage/adapters/fs/index.js (hunk 3 above)
Comment (wmertens) [inline lib/storage/adapters/fs/index.js]: Right, so adding some useful debug calls like when files are added and removed would be nice :)

File: lib/storage/index.js (hunk 4 above)
Comment (wmertens) [inline lib/storage/index.js]: Yes, I'll make an issue to track it.

Comment (wmertens) [inline lib/storage/index.js]: @JedWatson something like https://github.com/keystonejs/keystone/issues/3163?

**Comment by wmertens** [comment]
How would an Image type be made using this? Provide it with a File as a subfield and storing extra values for dimensions and thumbnails, or create from scratch with the storage lib?

**Comment by JedWatson** [comment]
New field type from scratch, using Storage the same way File does. That's the plan, at least. It would probably add some opinionated schema of its own on top of that provided by the Storage Adapter used (width, height, alt ...?) which we don't do in File. So we'll probably want to nest the data provided by the Storage API so those can never conflict.

e.g. Image would store:

```js
{
  src: {
    filename: 'wmertens.jpg',
    size: '42378',
    url: '/images/uploaded/wmertens.jpg',
    /* ... etc. file data as returned by the Storage Adapter */
  },
  width: 800,
  height: 625,
  alt: 'Wout!'
}
```

**Comment by JedWatson** [comment]
Riffing on that, we could implement a way to generate multiple image sizes (e.g. original, large, thumbnail) all of which contain their own file managed by Storage...

**Comment by JedWatson** [comment]
@webteckie see [model-transform](http://npmjs.com/package/model-transform)

**Comment by wmertens** [comment]
Yes, and what if the storage adapter is Cloudinary? Then the dimensions are known by Cloudinary and derivations are created by them… Just copy those values into the Image data?

**Comment by wmertens** [comment]
Right now we do multiple sizes by a separate process that autogenerates files in `_low`, `_mid` etc. Having it be part of Image is much nicer of course.

**Comment by JedWatson** [comment]
I'm planning to keep the Cloudinary fields totally separate because they have totally unique schema and behaviour.

**Comment by JedWatson** [comment]
Back to the spec above, I'm trying to get my head around the implications of what @josephg has suggested in terms of making `filename` something that the `FS` Adapter implements in the schema. There's some functionality in `Storage` for generating the filename which has to move, which is fine, but it leaves us without something we can rely on in terms of displaying the value actually stored, for generic Admin UI purposes.

Given the use case of something like a blob store which doesn't actually _have_ filenames, I think it does make sense to put it in the adapter. That leaves us needing to expose some instructions from the Adapter to the Admin UI on how to display an identifier for the File field value though. Ugh. We can't assume _anything_ from the UI perspective, if that's optional.

I'm going round in circles in my head here trying to work out which way to go on that, does anybody else have any input?

**Comment by wmertens** [comment]
I think the adapter should return the UUID for the file in a way that makes sense to the admin. For FS, that would be the path. For others, whatever.

Storage only knows about the original name and the displayUuid, and the adapter can maintain whatever values it needs for storing/retrieving the file.

**Comment by SharadKumar** [comment]
Instead of storing transformed images, Cloudinary type pattern of conversion and caching on the fly usually works out better. Not everyone wants to use Cloudinary.

So it might be good to add a middleware such as https://github.com/asilvas/node-image-steam, to mostly provide a good and fast (sharp lib) alternate to Cloudinary... use S3 or local storage. 

"Resizing an image is typically 4x-5x faster than using the quickest ImageMagick and GraphicsMagick settings." - Sharp.

**Comment by josephg** [comment]
@wmertens Yeah thats exactly what molomby and I said. There's two problems:

- Adapters like S3 have their primary key split between two fields (`bucket` and `id`). We could join them, but its a little bit icky
- The name of the primary key field changes depending on the adapter. Is it `filename` or `id` or `key` or just `base64data`?

**Comment by josephg** [comment]
@JedWatson I still think the admin UI should display the `displayname` field. If that doesn't exist in the schema, use `filename` or `originalname`. Then people can customize their own display name for files in the admin UI if they want (it should be configurable via a `generateDisplayName(file)` storage option). If they don't, we'll fall back to filename and originalname. Some people won't have any of those fields configured, and if so so be it. I mean, my database might literally just store the inlined base64 encoded file data. If so, what would you display in the admin UI? Nothing. And thats fine.

**Comment by JedWatson** [comment]
@josephg after further discussion on Slack with @wmertens I landed on adding a `nameProperty` property to adapters. For FS, it would be `FS.nameProperty = 'filename'`. This will get passed up to the React components in the Admin UI, which means each adapter can specify what it wants to display.

So for another adapter, it could be `displayName`. But since FS already has `filename` and that's what we want to display, this means we don't need to duplicate the value in the schema, and it's only a small amount of added complexity for the UI.

The limitations on the first version are that it's not a concatenation of multiple stored properties, and it's not a runtime function, but we can add those capabilities later if we want. I think this will be sufficient and flexible enough for now.

**Comment by JedWatson** [comment]
@SharadKumar awesome, I've been looking for something like that! I've seen sharp before and had it in mind for generating the cached images but this looks way better. Nice find!

**Comment by wmertens** [comment]
@sharadkumar, that's really nice! I wish I had known about it when I was
writing that resize-on-disk code... My only reservation about it is that it
won't fit with the access control that I built, which takes a path like
/files/listId/itemId/fileId/originalName, and sends the file if you have
access to it. This project expects to be the one that listens to the
request...

 @jedwatson, that reminds me, to have a generic access layer like that in
Keystone, you need a fileId. Thoughts? Right now I'm going through the
ˋattachmentsˋ array field looking for the file with filename fileId.

**Comment by JedWatson** [comment]
Storage and FS are now working, and have been tested in the `keystone-test-project`'s API Test UI. The Admin UI Components need some work (they're just including the LocalFile components right now) but I'm going to merge this into master so I can get this and the `fields-explorer` branch cleaned up - it's a mess at the moment, and I can't easily fix this up (including the linter errors) without conflicts otherwise.

Nearly there! :tada:

**Comment by mxstbr** [comment]
:tada:

**Comment by josephg** [comment]
Yaay!

I also really like that solution around `displayName`. 👍  great stuff.
